src/irmin-pack/IO.ml

```diff
@@ -245,53 +274,115 @@ module Unix : S = struct
           v ~offset:0L ~version:current_version ~generation:0L raw)
         else
           let version = Raw.Version.get raw in
-          match version_of_bin version with
+          match Version.of_bin version with
           | Some `V1 ->
               Log.debug (fun l -> l "[%s] file exists in V1" file);
               let offset = Raw.Offset.get raw in
               v ~offset ~version:`V1 ~generation:0L raw
           | Some `V2 ->
               let { Raw.Header.offset; generation; _ } = Raw.Header.get raw in
               v ~offset ~version:`V2 ~generation raw
-          | None ->
-              let pp_full_version ppf v =
-                Fmt.pf ppf "%a (%S)" pp_version v (bin_of_version v)
-              in
-              Fmt.failwith "invalid version: got %S, expecting %a" version
-                (Fmt.Dump.list pp_full_version)
-                (List.map fst versions))
+          | None -> Version.raise_invalid version)
 
   let close t = Raw.close t.raw
+
+  (* From a given offset in [src], transfer all data to [dst] (starting at
+     [dst_off]). *)
+  let transfer_all ~progress ~src ~src_off ~dst ~dst_off =
+    let ( + ) a b = Int64.(add a (of_int b)) in
+    let buf_len = 4096 in
+    let buf = Bytes.create buf_len in
+    let rec inner ~src_off ~dst_off =
+      match Raw.unsafe_read src ~off:src_off ~len:buf_len buf with
+      | 0 -> ()
+      | read ->
+          assert (read <= buf_len);
+          let to_write = if read < buf_len then Bytes.sub buf 0 read else buf in
+          let () =
+            Raw.unsafe_write dst ~off:dst_off (Bytes.unsafe_to_string to_write)
+          in
+          progress (Int64.of_int read);
+          (inner [@tailcall]) ~src_off:(src_off + read) ~dst_off:(dst_off + read)
+    in
+    inner ~src_off ~dst_off
+
+  let upgrade ~src ~dst:(dst_path, dst_v) ~progress =
+    let src_v =
+      let v_bin = Raw.Version.get src.raw in
+      match Version.of_bin v_bin with
+      | None -> Fmt.failwith "Could not parse version string `%s'" v_bin
+      | Some v -> v
+    in
+    let src_offset = Raw.Offset.get src.raw in
+    match (src_v, dst_v) with
+    | `V1, `V2 ->
+        Log.debug (fun m ->
+            m "[%s] Performing migration: %a → %a"
+              (Filename.basename src.file)
+              pp_version `V1 pp_version `V2);
+        let dst =
+          (* Note: all V1 files implicitly have [generation = 0], since it
+             is not possible to [clear] them. *)
+          raw dst_path
+            ~flags:[ Unix.O_CREAT; O_WRONLY; O_APPEND ]
```

Comment:
does it really matter that you open the file on `O_APPEND`?

Reply:
I'm not actually sure. I vaguely thought that using `O_APPEND` would speed up `pwrite` by avoiding the positioning step, but the documentation only mentions `write`:

```
       O_APPEND
              The file is opened in append mode.  Before each write(2), the
              file offset is positioned at the end of the file, as if with
              lseek(2).  The modification of the file offset and the write
              operation are performed as a single atomic step.
```

Happy to take it out.

Reply:
I'd just like to avoid (if possible) exposing the POSIX open flags in the IO API.

Reply:
Removed :+1: